File: src/ui/scripts/common/download-config.ts
interface IPCDownloadInputConfig {
    url:           string;
    is_playlist:   boolean;
    download_type: string;
}

interface IPCDownloadSettingsConfig {
    format:                         string;
    format_type:                    string;

    trim_enable:                    boolean;
    trim_from_start_enable:         boolean;
    trim_start:                     string;
    trim_to_end_enable:             boolean;
    trim_end:                       string;

    size_change_enable:             boolean;
    size_change_width:              string;
    size_change_height:             string;

    fps_change_enable:              boolean;
    fps_change_framerate:           string;

    vbr_bitrate:                    string;
    abr_bitrate:                    string;

    custom_ffmpeg_arguments_enable: boolean;
    custom_ffmpeg_arguments:        string[];

    custom_ytdlp_arguments_enable:  boolean;
    custom_ytdlp_arguments:         string[];
}

interface IPCDownloadOutputConfig {
    name: string;
    path: string;
}

export interface IPCDownloadConfig {
    input:    IPCDownloadInputConfig;
    settings: IPCDownloadSettingsConfig;
    output:   IPCDownloadOutputConfig;
}

export function createIPCDownloadConfig() {
    const config: IPCDownloadConfig = {
        input: {
            url:           "",
            is_playlist:   false,
            download_type: ""
        },

        settings: {
            format:                         "",
            format_type:                    "",

            trim_enable:                    false,
            trim_from_start_enable:         false,
            trim_start:                     "",
            trim_to_end_enable:             false,
            trim_end:                       "",

            size_change_enable:             false,
            size_change_width:              "",
            size_change_height:             "",

            fps_change_enable:              false,
            fps_change_framerate:           "",

            vbr_bitrate:                    "",
            abr_bitrate:                    "",

            custom_ytdlp_arguments_enable:  false,
            custom_ytdlp_arguments:         [],
            custom_ffmpeg_arguments_enable: false,
            custom_ffmpeg_arguments:        []
        },

        output: {
            name: "",
            path: ""
        }
    };

    return config;
}

File: src/app/src/config/download_config.rs
use serde::{Deserialize, Serialize};

use crate::utils::string::clean_string_vector;

#[derive(Serialize, Deserialize, Clone, Default)]
pub struct IPCDownloadInputConfig {
    pub url: String,
    pub is_playlist: bool,
    pub download_type: String,
}

#[derive(Serialize, Deserialize, Clone, Default)]
pub struct IPCDownloadSettingsConfig {
    pub format: String,
    pub format_type: String,

    pub trim_enable: bool,
    pub trim_from_start_enable: bool,
    pub trim_start: String,
    pub trim_to_end_enable: bool,
    pub trim_end: String,

    pub size_change_enable: bool,
    pub size_change_width: String,
    pub size_change_height: String,

    pub fps_change_enable: bool,
    pub fps_change_framerate: String,

    pub vbr_set_bitrate_enable: bool,
    pub vbr_set_bitrate: String,
    pub abr_set_bitrate_enable: bool,
    pub abr_set_bitrate: String,

    pub custom_ytdlp_arguments_enable: bool,
    pub custom_ytdlp_arguments: Vec<String>,

    pub custom_ffmpeg_arguments_enable: bool,
    pub custom_ffmpeg_arguments: Vec<String>,
}

#[derive(Serialize, Deserialize, Clone, Default)]
pub struct IPCDownloadOutputConfig {
    pub name: String,
    pub path: String,
}

#[derive(Serialize, Deserialize, Clone, Default)]
pub struct IPCDownloadConfig {
    pub valid: bool,

    pub input: IPCDownloadInputConfig,
    pub settings: IPCDownloadSettingsConfig,
    pub output: IPCDownloadOutputConfig,
}

impl IPCDownloadConfig {
    pub fn purify(&mut self) {
        clean_string_vector(&mut self.settings.custom_ytdlp_arguments);
        clean_string_vector(&mut self.settings.custom_ffmpeg_arguments);
    }
}
